File: packages/core/src/drawings/lz-string.ts
/**
 * The base64 form of LZ-String, vendored from lz-string 1.5.0 (https://github.com/pieroxy/lz-string)
 * because `@ddl/core` has no dependencies. The Obsidian Excalidraw plugin stores drawings as
 * `compressToBase64` output (its `compressed-json` blocks), so this must stay bit-compatible with
 * the library. Differences: typed, the bit loops are shared helpers, invalid input returns null
 * instead of a garbage string, and decompression stops at `MAX_OUTPUT_CHARS`.
 *
 * MIT License
 *
 * Copyright (c) 2013 pieroxy
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const BASE64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";

/**
 * LZW output can grow quadratically with crafted input; a real drawing decompresses to a few MB.
 * Past this many UTF-16 code units the input is treated as invalid.
 */
const MAX_OUTPUT_CHARS = 64 * 1024 * 1024;

let base64Values: Int8Array | undefined;

/** The library maps characters outside the alphabet to `undefined`, which its bit math reads as 0. */
function base64Value(code: number): number {
  if (!base64Values) {
    base64Values = new Int8Array(128);
    for (let i = 0; i < BASE64_ALPHABET.length; i++)
      base64Values[BASE64_ALPHABET.charCodeAt(i)] = i;
  }
  return code < 128 ? base64Values[code]! : 0;
}

/** `LZString.compressToBase64`. */
export function compressToBase64(input: string): string {
  const out = compress(input, 6, (value) => BASE64_ALPHABET.charAt(value));
  switch (out.length % 4) {
    case 1:
      return `${out}===`;
    case 2:
      return `${out}==`;
    case 3:
      return `${out}=`;
    default:
      return out;
  }
}

/**
 * `LZString.decompressFromBase64`: the original text, `""` for an empty or truncated stream, or
 * null when `input` isn't LZ-String base64. Callers strip line breaks first.
 */
export function decompressFromBase64(input: string): string | null {
  if (input === "") return null;
  return decompress(input.length, 32, (index) => base64Value(input.charCodeAt(index)));
}

function compress(
  uncompressed: string,
  bitsPerChar: number,
  charFor: (value: number) => string,
): string {
  const dictionary = new Map<string, number>();
  const toCreate = new Set<string>();
  const data: string[] = [];
  let w = "";
  let enlargeIn = 2;
  let dictSize = 3;
  let numBits = 2;
  let bits = 0;
  let position = 0;

  const write = (count: number, value: number) => {
    for (let i = 0; i < count; i++) {
      bits = (bits << 1) | (value & 1);
      if (position === bitsPerChar - 1) {
        position = 0;
        data.push(charFor(bits));
        bits = 0;
      } else {
        position++;
      }
      value >>= 1;
    }
  };
  const countDown = () => {
    enlargeIn--;
    if (enlargeIn === 0) {
      enlargeIn = 2 ** numBits;
      numBits++;
    }
  };
  const emit = (phrase: string) => {
    if (toCreate.has(phrase)) {
      const code = phrase.charCodeAt(0);
      if (code < 256) {
        write(numBits, 0);
        write(8, code);
      } else {
        write(numBits, 1);
        write(16, code);
      }
      countDown();
      toCreate.delete(phrase);
    } else {
      write(numBits, dictionary.get(phrase)!);
    }
    countDown();
  };

  for (let i = 0; i < uncompressed.length; i++) {
    const c = uncompressed.charAt(i);
    if (!dictionary.has(c)) {
      dictionary.set(c, dictSize++);
      toCreate.add(c);
    }
    const wc = w + c;
    if (dictionary.has(wc)) {
      w = wc;
    } else {
      emit(w);
      dictionary.set(wc, dictSize++);
      w = c;
    }
  }
  if (w !== "") emit(w);
  write(numBits, 2);
  for (;;) {
    bits <<= 1;
    if (position === bitsPerChar - 1) {
      data.push(charFor(bits));
      break;
    }
    position++;
  }
  return data.join("");
}

function decompress(
  length: number,
  resetValue: number,
  next: (index: number) => number,
): string | null {
  const dictionary: string[] = [];
  let enlargeIn = 4;
  let dictSize = 4;
  let numBits = 3;
  let value = next(0);
  let position = resetValue;
  let index = 1;
  let outputChars = 0;

  const read = (count: number): number => {
    let bits = 0;
    for (let power = 1, max = 2 ** count; power !== max; power *= 2) {
      const bit = value & position;
      position >>= 1;
      if (position === 0) {
        position = resetValue;
        value = next(index++);
      }
      if (bit > 0) bits += power;
    }
    return bits;
  };

  let first: string;
  switch (read(2)) {
    case 0:
      first = String.fromCharCode(read(8));
      break;
    case 1:
      first = String.fromCharCode(read(16));
      break;
    case 2:
      return "";
    default:
      return null;
  }
  dictionary[3] = first;
  let w = first;
  const result = [first];
  outputChars = first.length;
  for (;;) {
    if (index > length) return "";
    let code = read(numBits);
    switch (code) {
      case 0:
        dictionary[dictSize++] = String.fromCharCode(read(8));
        code = dictSize - 1;
        enlargeIn--;
        break;
      case 1:
        dictionary[dictSize++] = String.fromCharCode(read(16));
        code = dictSize - 1;
        enlargeIn--;
        break;
      case 2:
        return result.join("");
    }
    if (enlargeIn === 0) {
      enlargeIn = 2 ** numBits;
      numBits++;
    }
    let entry: string;
    const known = dictionary[code];
    if (known) entry = known;
    else if (code === dictSize) entry = w + w.charAt(0);
    else return null;
    outputChars += entry.length;
    if (outputChars > MAX_OUTPUT_CHARS) return null;
    result.push(entry);
    dictionary[dictSize++] = w + entry.charAt(0);
    enlargeIn--;
    w = entry;
    if (enlargeIn === 0) {
      enlargeIn = 2 ** numBits;
      numBits++;
    }
  }
}

File: apps/macos/Packages/DailyDoListDrawing/Sources/DailyDoListDrawingModel/File/LZString.swift
// A port of LZ-String 1.5.0's base64 codec (lz-string/libs/lz-string.js), which the Obsidian
// Excalidraw plugin uses for `compressed-json` drawings.
//
// MIT License
//
// Copyright (c) 2013 pieroxy
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Foundation

/// LZ-String's `compressToBase64` and `decompressFromBase64`, on UTF-16 code units like the
/// JavaScript original, so output matches it character for character.
public enum LZString {
  static let base64Alphabet = Array(
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=".utf16)

  static let base64Values: [UInt16: Int] = {
    var values: [UInt16: Int] = [:]
    for (index, unit) in base64Alphabet.enumerated() { values[unit] = index }
    return values
  }()

  public static func compressToBase64(_ input: String) -> String {
    let units = compress(Array(input.utf16), bitsPerChar: 6) { base64Alphabet[$0] }
    var result = String(decoding: units, as: UTF16.self)
    switch units.count % 4 {
    case 1: result += "==="
    case 2: result += "=="
    case 3: result += "="
    default: break
    }
    return result
  }

  /// nil when the input isn't valid LZ-String data (JavaScript returns null or "").
  public static func decompressFromBase64(_ input: String) -> String? {
    let units = Array(input.utf16)
    guard !units.isEmpty else { return nil }
    guard
      let output = decompress(
        length: units.count, resetValue: 32,
        next: { index in index < units.count ? (base64Values[units[index]] ?? 0) : 0 })
    else { return nil }
    return String(decoding: output, as: UTF16.self)
  }

  // MARK: Compression

  private struct BitWriter {
    let bitsPerChar: Int
    let character: (Int) -> UInt16
    var data: [UInt16] = []
    var value = 0
    var position = 0

    mutating func write(_ bits: Int, count: Int) {
      var bits = bits
      for _ in 0..<count {
        value = (value << 1) | (bits & 1)
        if position == bitsPerChar - 1 {
          position = 0
          data.append(character(value))
          value = 0
        } else {
          position += 1
        }
        bits >>= 1
      }
    }

    /// The "1 then zeros" marker of a 16-bit literal, written high bit first like the original.
    mutating func writeMarker(_ marker: Int, count: Int) {
      var marker = marker
      for _ in 0..<count {
        value = (value << 1) | marker
        if position == bitsPerChar - 1 {
          position = 0
          data.append(character(value))
          value = 0
        } else {
          position += 1
        }
        marker = 0
      }
    }

    mutating func flush() {
      while true {
        value <<= 1
        if position == bitsPerChar - 1 {
          data.append(character(value))
          break
        }
        position += 1
      }
    }
  }

  static func compress(_ input: [UInt16], bitsPerChar: Int, character: @escaping (Int) -> UInt16)
    -> [UInt16]
  {
    var singles: [UInt16: Int] = [:]
    var pendingSingles = Set<UInt16>()
    var pairs: [UInt64: Int] = [:]
    var enlargeIn = 2
    var dictSize = 3
    var numBits = 2
    var writer = BitWriter(bitsPerChar: bitsPerChar, character: character)
    // w: nothing yet, a single character, or a longer dictionary entry (its code).
    var wCode: Int?
    var wSingle: UInt16?

    func emitW() {
      if let single = wSingle, pendingSingles.contains(single) {
        if single < 256 {
          writer.write(0, count: numBits)
          writer.write(Int(single), count: 8)
        } else {
          writer.writeMarker(1, count: numBits)
          writer.write(Int(single), count: 16)
        }
        enlargeIn -= 1
        if enlargeIn == 0 {
          enlargeIn = 1 << numBits
          numBits += 1
        }
        pendingSingles.remove(single)
      } else if let wCode {
        writer.write(wCode, count: numBits)
      }
      enlargeIn -= 1
      if enlargeIn == 0 {
        enlargeIn = 1 << numBits
        numBits += 1
      }
    }

    for c in input {
      if singles[c] == nil {
        singles[c] = dictSize
        dictSize += 1
        pendingSingles.insert(c)
      }
      guard let current = wCode else {
        wCode = singles[c]
        wSingle = c
        continue
      }
      let key = UInt64(current) << 16 | UInt64(c)
      if let code = pairs[key] {
        wCode = code
        wSingle = nil
        continue
      }
      emitW()
      pairs[key] = dictSize
      dictSize += 1
      wCode = singles[c]
      wSingle = c
    }
    if wCode != nil { emitW() }
    writer.write(2, count: numBits)
    writer.flush()
    return writer.data
  }

  // MARK: Decompression

  static func decompress(length: Int, resetValue: Int, next: (Int) -> Int) -> [UInt16]? {
    var dictionary: [[UInt16]] = [[0], [1], [2]]
    var enlargeIn = 4
    var numBits = 3
    var result: [UInt16] = []
    var value = next(0)
    var position = resetValue
    var index = 1

    func readBits(_ count: Int) -> Int {
      var bits = 0
      var power = 1
      let maxPower = 1 << count
      while power != maxPower {
        let bit = value & position
        position >>= 1
        if position == 0 {
          position = resetValue
          value = next(index)
          index += 1
        }
        bits |= (bit > 0 ? 1 : 0) * power
        power <<= 1
      }
      return bits
    }

    let c: [UInt16]
    switch readBits(2) {
    case 0: c = [UInt16(readBits(8))]
    case 1: c = [UInt16(readBits(16))]
    default: return []
    }
    dictionary.append(c)
    var w = c
    result.append(contentsOf: c)
    while true {
      if index > length { return [] }
      var code = readBits(numBits)
      switch code {
      case 0:
        dictionary.append([UInt16(readBits(8))])
        code = dictionary.count - 1
        enlargeIn -= 1
      case 1:
        dictionary.append([UInt16(readBits(16))])
        code = dictionary.count - 1
        enlargeIn -= 1
      case 2:
        return result
      default:
        break
      }
      if enlargeIn == 0 {
        enlargeIn = 1 << numBits
        numBits += 1
      }
      let entry: [UInt16]
      if code < dictionary.count {
        entry = dictionary[code]
      } else if code == dictionary.count {
        entry = w + [w[0]]
      } else {
        return nil
      }
      result.append(contentsOf: entry)
      dictionary.append(w + [entry[0]])
      enlargeIn -= 1
      w = entry
      if enlargeIn == 0 {
        enlargeIn = 1 << numBits
        numBits += 1
      }
    }
  }
}
